### DISCUSSION TITLE
How does grdb decode numbers to enums?

### DISCUSSION THREAD

**Discussion by mikechambers** [groue/GRDB.swift#1447] [Q&A]
When GRDB is trying to decode an Int into an enum, does it just call init(rawValue)?

i.e. if I have this:

```
enum Standing: Int32, Decodable {
	case victory = 0
	case defeat = 1
	case unknown = 255
	
	init(rawValue:Int32) {
		if rawValue == 0 {
			self = .victory
		} else {
			self = .defeat
		}
	}
}
```

and standing is stored as an Int32 in the DB, will it call init(rawValue) to convert to the enum?

**Comment by groue** [comment]
Hello @mikechambers,

Well, it depends.

When decoding a `Standing` property embedded in a `Decodable` record type, it is the `Decodable` facet of `Standing` that performs the decoding, with `init(from decoder: Decoder)`. This one is synthesized by the compiler, and I haven't checked if it delegates to `init(rawValue:)` or not. You will perform your own check if necessary.

It is also possible to add an explicit conformance to [`DatabaseValueConvertible`](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/databasevalueconvertible). This conformance takes precedence over the eventual `Decodable` conformance, and makes it possible to precisely control the decoding of database values.

By default, `DatabaseValueConvertible` calls `init(rawValue:)` for `RawRepresentable` types like `Standing`:

```swift
// Makes sure Standing values loaded from the
// database are decoded with init(rawValue:).
extension Standing: DatabaseValueConvertible { }
```

**Reply by mikechambers** [reply]
fyi:

>This one is synthesized by the compiler, and I haven't checked if it delegates to init(rawValue:) or not. You will perform your own check if necessary.

It appears it does call init(rawValue) under the hood

thanks for the input
